File: src/dmp_base/TransformationSystem.cpp
#include "dmp/dmp_base/TransformationSystem.h"

#include <memory>

namespace dmp {

TransformationSystem::TransformationSystem()
    : dmp_num_dimensions(0),
      is_started(false),
      is_using_coupling_term_at_dimension(std::vector<bool>()),
      tau_sys(NULL),
      func_approx(nullptr),
      transform_coupling(nullptr),
      start_state(nullptr),
      current_state(nullptr),
      current_velocity_state(nullptr),
      goal_sys(nullptr),
      logged_dmp_variables(nullptr),
      rt_assertor(nullptr) {}

TransformationSystem::TransformationSystem(
    uint dmp_num_dimensions_init, TauSystem* tau_system,
    CanonicalSystem* canonical_system,
    FunctionApproximator* function_approximator,
    LoggedDMPVariables* logged_dmp_vars, RealTimeAssertor* real_time_assertor,
    std::unique_ptr<DMPState> start_dmpstate,
    std::unique_ptr<DMPState> current_dmpstate,
    std::unique_ptr<DMPState> current_velocity_dmpstate,
    std::unique_ptr<GoalSystem> goal_system,
    std::vector<TransformCoupling*>* transform_couplers)
    : dmp_num_dimensions(dmp_num_dimensions_init),
      is_started(false),
      is_using_coupling_term_at_dimension(
          std::vector<bool>(dmp_num_dimensions_init, true)),
      tau_sys(tau_system),
      func_approx(function_approximator),
      logged_dmp_variables(logged_dmp_vars),
      rt_assertor(real_time_assertor),
      transform_coupling(transform_couplers) {
  if (start_dmpstate) {
    start_state = std::move(start_dmpstate);
  } else {
    start_state =
        std::make_unique<DMPState>(dmp_num_dimensions_init, real_time_assertor);
  }

  if (current_dmpstate) {
    current_state = std::move(current_dmpstate);
  } else {
    current_state =
        std::make_unique<DMPState>(dmp_num_dimensions_init, real_time_assertor);
  }

  if (current_velocity_dmpstate) {
    current_velocity_state = std::move(current_velocity_dmpstate);
  } else {
    current_velocity_state =
        std::make_unique<DMPState>(dmp_num_dimensions_init, real_time_assertor);
  }

  if (goal_system) {
    goal_sys = std::move(goal_system);
  } else {
    goal_sys = std::make_unique<GoalSystem>(dmp_num_dimensions_init, tau_system,
                                            real_time_assertor);
  }
}

bool TransformationSystem::isValid() {
  if (rt_assert((rt_assert(dmp_num_dimensions > 0)) &&
                (rt_assert(dmp_num_dimensions <= MAX_DMP_NUM_DIMENSIONS)) &&
                (rt_assert(is_using_coupling_term_at_dimension.size() ==
                           dmp_num_dimensions))) == false) {
    return false;
  }
  if (rt_assert((rt_assert(tau_sys != NULL)) &&
                (rt_assert(func_approx != NULL)) &&
                (rt_assert(goal_sys != NULL))) == false) {
    return false;
  }
  if (rt_assert((rt_assert(tau_sys->isValid())) &&
                (rt_assert(func_approx->isValid())) &&
                (rt_assert(goal_sys->isValid()))) == false) {
    return false;
  }
  if (rt_assert((rt_assert(tau_sys == goal_sys->getTauSystemPointer())) &&
                (rt_assert(tau_sys == func_approx->getCanonicalSystemPointer()
                                          ->getTauSystemPointer()))) == false) {
    return false;
  }
  if (rt_assert((rt_assert(start_state != NULL)) &&
                (rt_assert(current_state != NULL)) &&
                (rt_assert(current_velocity_state != NULL))) == false) {
    return false;
  }
  if (rt_assert((rt_assert(start_state->isValid())) &&
                (rt_assert(current_state->isValid())) &&
                (rt_assert(current_velocity_state->isValid()))) == false) {
    return false;
  }
  if (rt_assert(
          (rt_assert(func_approx->getDMPNumDimensions() ==
                     dmp_num_dimensions)) &&
          (rt_assert(goal_sys->getDMPNumDimensions() == dmp_num_dimensions)) &&
          (rt_assert(start_state->getDMPNumDimensions() ==
                     dmp_num_dimensions)) &&
          (rt_assert(current_state->getDMPNumDimensions() ==
                     dmp_num_dimensions)) &&
          (rt_assert(current_velocity_state->getDMPNumDimensions() ==
                     dmp_num_dimensions))) == false) {
    return false;
  }
  return true;
}

bool TransformationSystem::getCouplingTerm(VectorN& accumulated_ct_acc,
                                           VectorN& accumulated_ct_vel) {
  // input checking
  if (rt_assert(accumulated_ct_acc.rows() == dmp_num_dimensions) == false) {
    return false;
  }
  if (rt_assert(accumulated_ct_vel.rows() == dmp_num_dimensions) == false) {
    return false;
  }

  accumulated_ct_acc = ZeroVectorN(dmp_num_dimensions);
  accumulated_ct_vel = ZeroVectorN(dmp_num_dimensions);

  if (transform_coupling) {
    for (uint i = 0; i < (*transform_coupling).size(); ++i) {
      if ((*transform_coupling)[i] == NULL) {
        continue;
      }

      VectorN ct_acc(dmp_num_dimensions);
      VectorN ct_vel(dmp_num_dimensions);
      ct_acc = ZeroVectorN(dmp_num_dimensions);
      ct_vel = ZeroVectorN(dmp_num_dimensions);
      if (rt_assert((*transform_coupling)[i]->getValue(ct_acc, ct_vel)) ==
          false) {
        return false;
      }
      if (rt_assert((containsNaN(ct_acc) || containsNaN(ct_vel)) == false) ==
          false) {
        return false;
      }

      accumulated_ct_acc += ct_acc;
      accumulated_ct_vel += ct_vel;
    }
  }

  return true;
}

uint TransformationSystem::getDMPNumDimensions() { return dmp_num_dimensions; }

uint TransformationSystem::getModelSize() {
  return func_approx->getModelSize();
}

DMPState TransformationSystem::getStartState() { return (*start_state); }

bool TransformationSystem::setStartState(const DMPState& new_start_state) {
  // pre-conditions checking
  if (rt_assert(this->isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(new_start_state.isValid()) == false) {
    return false;
  }
  if (rt_assert(start_state->getDMPNumDimensions() ==
                new_start_state.getDMPNumDimensions()) == false) {
    return false;
  }

  *start_state = new_start_state;

  // post-condition(s) checking
  return (rt_assert(this->isValid()));
}

DMPState TransformationSystem::getCurrentState() { return (*current_state); }

bool TransformationSystem::setCurrentState(const DMPState& new_current_state) {
  // pre-conditions checking
  if (rt_assert(this->isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(new_current_state.isValid()) == false) {
    return false;
  }
  if (rt_assert(current_state->getDMPNumDimensions() ==
                new_current_state.getDMPNumDimensions()) == false) {
    return false;
  }

  *current_state = new_current_state;

  if (rt_assert(this->updateCurrentVelocityStateFromCurrentState()) == false) {
    return false;
  }

  // post-condition(s) checking
  return (rt_assert(this->isValid()));
}

bool TransformationSystem::updateCurrentVelocityStateFromCurrentState() {
  // pre-conditions checking
  if (rt_assert(this->isValid()) == false) {
    return false;
  }

  double tau;
  if (rt_assert(tau_sys->getTauRelative(tau)) == false) {
    return false;
  }

  VectorN xd = current_state->getXd();
  VectorN xdd = current_state->getXdd();

  VectorN v = tau * xd;
  VectorN vd = tau * xdd;

  if (rt_assert(current_velocity_state->setX(v)) == false) {
    return false;
  }
  if (rt_assert(current_velocity_state->setXd(vd)) == false) {
    return false;
  }

  // post-conditions checking
  return (rt_assert(this->isValid()));
}

DMPState TransformationSystem::getCurrentGoalState() {
  return (goal_sys->getCurrentGoalState());
}

bool TransformationSystem::setCurrentGoalState(
    const DMPState& new_current_goal_state) {
  return (rt_assert(goal_sys->setCurrentGoalState(new_current_goal_state)));
}

bool TransformationSystem::updateCurrentGoalState(const double& dt) {
  return (rt_assert(goal_sys->updateCurrentGoalState(dt)));
}

VectorN TransformationSystem::getSteadyStateGoalPosition() {
  return (goal_sys->getSteadyStateGoalPosition());
}

bool TransformationSystem::setSteadyStateGoalPosition(const VectorN& new_G) {
  return (rt_assert(goal_sys->setSteadyStateGoalPosition(new_G)));
}

TauSystem* TransformationSystem::getTauSystemPointer() { return tau_sys; }

CanonicalSystem* TransformationSystem::getCanonicalSystemPointer() {
  return func_approx->getCanonicalSystemPointer();
}

FunctionApproximator* TransformationSystem::getFuncApproxPointer() {
  return func_approx;
}

bool TransformationSystem::setCouplingTermUsagePerDimensions(
    const std::vector<bool>& is_using_coupling_term_at_dimension_init) {
  if (rt_assert(is_using_coupling_term_at_dimension_init.size() ==
                dmp_num_dimensions) == false) {
    return false;
  }

  is_using_coupling_term_at_dimension =
      is_using_coupling_term_at_dimension_init;

  return true;
}

TransformationSystem::~TransformationSystem() {}

}  // namespace dmp

File: python/dmp_base/TransformationSystem.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
sys.path.append(os.path.join(os.path.dirname(__file__), "../dmp_goal_system/"))
sys.path.append(os.path.join(os.path.dirname(__file__), "../dmp_param/"))
sys.path.append(os.path.join(os.path.dirname(__file__), "../dmp_state/"))
sys.path.append(
    os.path.join(os.path.dirname(__file__), "../dmp_coupling/base/"))
from TauSystem import *
from CanonicalSystem import *
from FunctionApproximator import *
from TransformCoupling import *
from DMPState import *
from GoalSystem import *

MIN_FABS_AMPLITUDE = 1.e-3


class TransformationSystem:
  "Base class for transformation systems of DMPs."

  def __init__(self,
               dmp_num_dimensions_init,
               canonical_system,
               function_approximator,
               start_dmpstate=None,
               current_dmpstate=None,
               current_velocity_dmpstate=None,
               goal_system=None,
               transform_couplers_list=[],
               name=""):
    self.name = name
    self.dmp_num_dimensions = dmp_num_dimensions_init
    self.canonical_sys = canonical_system
    self.tau_sys = self.canonical_sys.tau_sys
    self.func_approx = function_approximator
    self.transform_couplers_list = transform_couplers_list
    self.is_using_coupling_term_at_dimension = [True] * self.dmp_num_dimensions
    self.is_started = False
    if (start_dmpstate != None):
      self.start_state = start_dmpstate
    else:
      self.start_state = DMPState(np.zeros((self.dmp_num_dimensions, 1)))
    if (current_dmpstate != None):
      self.current_state = current_dmpstate
    else:
      self.current_state = DMPState(np.zeros((self.dmp_num_dimensions, 1)))
    if (current_velocity_dmpstate != None):
      self.current_velocity_state = current_velocity_dmpstate
    else:
      self.current_velocity_state = DMPState(
          np.zeros((self.dmp_num_dimensions, 1)))
    if (goal_system != None):
      self.goal_sys = goal_system
    else:
      self.goal_sys = GoalSystem(self.dmp_num_dimensions, self.tau_sys)

  def isValid(self):
    assert (self.dmp_num_dimensions > 0), "self.dmp_num_dimensions=" + str(
        self.dmp_num_dimensions) + " <= 0 (invalid!)"
    assert (len(
        self.is_using_coupling_term_at_dimension) == self.dmp_num_dimensions
           ), "len(self.is_using_coupling_term_at_dimension)=" + str(
               len(self.is_using_coupling_term_at_dimension)
           ) + " is mis-matched with self.dmp_num_dimensions=" + str(
               self.dmp_num_dimensions) + "!"
    assert (self.tau_sys != None)
    assert (self.canonical_sys != None)
    assert (self.func_approx != None)
    assert (self.goal_sys != None)
    assert (self.tau_sys.isValid())
    assert (self.canonical_sys.isValid())
    assert (self.func_approx.isValid())
    assert (self.goal_sys.isValid())
    assert (self.tau_sys == self.canonical_sys.tau_sys)
    assert (self.tau_sys == self.goal_sys.tau_sys)
    assert (self.start_state != None)
    assert (self.current_state != None)
    assert (self.current_velocity_state != None)
    assert (self.start_state.isValid())
    assert (self.current_state.isValid())
    assert (self.current_velocity_state.isValid())
    assert (self.func_approx.dmp_num_dimensions == self.dmp_num_dimensions
           ), "self.func_approx.dmp_num_dimensions=" + str(
               self.func_approx.dmp_num_dimensions
           ) + " is mis-matched with self.dmp_num_dimensions=" + str(
               self.dmp_num_dimensions) + "!"
    assert (self.goal_sys.dmp_num_dimensions == self.dmp_num_dimensions
           ), "self.goal_sys.dmp_num_dimensions=" + str(
               self.goal_sys.dmp_num_dimensions
           ) + " is mis-matched with self.dmp_num_dimensions=" + str(
               self.dmp_num_dimensions) + "!"
    assert (self.start_state.dmp_num_dimensions == self.dmp_num_dimensions
           ), "self.start_state.dmp_num_dimensions=" + str(
               self.start_state.dmp_num_dimensions
           ) + " is mis-matched with self.dmp_num_dimensions=" + str(
               self.dmp_num_dimensions) + "!"
    assert (self.current_state.dmp_num_dimensions == self.dmp_num_dimensions
           ), "self.current_state.dmp_num_dimensions=" + str(
               self.current_state.dmp_num_dimensions
           ) + " is mis-matched with self.dmp_num_dimensions=" + str(
               self.dmp_num_dimensions) + "!"
    assert (self.current_velocity_state.dmp_num_dimensions ==
            self.dmp_num_dimensions
           ), "self.current_velocity_state.dmp_num_dimensions=" + str(
               self.current_velocity_state.dmp_num_dimensions
           ) + " is mis-matched with self.dmp_num_dimensions=" + str(
               self.dmp_num_dimensions) + "!"
    return True

  def resetCouplingTerm(self):
    for transform_coupler_idx in range(len(self.transform_couplers_list)):
      if (self.transform_couplers_list[transform_coupler_idx] is not None):
        ret = self.transform_couplers_list[transform_coupler_idx].reset()
        if (ret == False):
          return False
    return True

  def getCouplingTerm(self):
    accumulated_ct_acc = np.zeros((self.dmp_num_dimensions, 1))
    accumulated_ct_vel = np.zeros((self.dmp_num_dimensions, 1))
    for transform_coupler_idx in range(len(self.transform_couplers_list)):
      if (self.transform_couplers_list[transform_coupler_idx] is not None):
        ct_acc, ct_vel = self.transform_couplers_list[
            transform_coupler_idx].getValue()
        assert (np.isnan(ct_acc).all() == False
               ), "ct_acc[" + str(transform_coupler_idx) + "] is NaN!"
        assert (np.isnan(ct_vel).all() == False
               ), "ct_vel[" + str(transform_coupler_idx) + "] is NaN!"
        accumulated_ct_acc = accumulated_ct_acc + ct_acc
        accumulated_ct_vel = accumulated_ct_vel + ct_vel
    assert (accumulated_ct_acc.shape == (self.dmp_num_dimensions, 1))
    assert (accumulated_ct_vel.shape == (self.dmp_num_dimensions, 1))
    return accumulated_ct_acc, accumulated_ct_vel

  def getStartState(self):
    return copy.copy(self.start_state)

  def setStartState(self, new_start_state):
    assert (self.isValid()), (
        "Pre-condition(s) checking is failed: this TransformationSystem is "
        "invalid!")
    assert (new_start_state.isValid())
    assert (
        self.start_state.dmp_num_dimensions == new_start_state
        .dmp_num_dimensions), "self.start_state.dmp_num_dimensions=" + str(
            self.start_state.dmp_num_dimensions
        ) + " is mis-matched with new_start_state.dmp_num_dimensions=" + str(
            new_start_state.dmp_num_dimensions) + "!"
    self.start_state = copy.copy(new_start_state)
    assert (self.isValid()), (
        "Post-condition(s) checking is failed: this TransformationSystem became"
        " invalid!")
    return None

  def getCurrentState(self):
    return copy.copy(self.current_state)

  def setCurrentState(self, new_current_state):
    assert (self.isValid()), (
        "Pre-condition(s) checking is failed: this TransformationSystem is "
        "invalid!")
    assert (new_current_state.isValid())
    assert (
        self.current_state.dmp_num_dimensions == new_current_state
        .dmp_num_dimensions), "self.current_state.dmp_num_dimensions=" + str(
            self.current_state.dmp_num_dimensions
        ) + " is mis-matched with new_current_state.dmp_num_dimensions=" + str(
            new_current_state.dmp_num_dimensions) + "!"
    self.current_state = copy.copy(new_current_state)
    self.updateCurrentVelocityStateFromCurrentState()
    assert (self.isValid()), (
        "Post-condition(s) checking is failed: this TransformationSystem became"
        " invalid!")
    return None

  def updateCurrentVelocityStateFromCurrentState(self):
    assert (self.isValid()), (
        "Pre-condition(s) checking is failed: this TransformationSystem is "
        "invalid!")

    tau = self.tau_sys.getTauRelative()
    Xd = self.current_state.getXd()
    Xdd = self.current_state.getXdd()
    V = tau * Xd
    Vd = tau * Xdd
    self.current_velocity_state.setX(V)
    self.current_velocity_state.setXd(Vd)

    assert (self.isValid()), (
        "Post-condition(s) checking is failed: this TransformationSystem became"
        " invalid!")
    return None

  def getCurrentGoalState(self):
    return self.goal_sys.getCurrentGoalState()

  def setCurrentGoalState(self, new_current_goal_state):
    return self.goal_sys.setCurrentGoalState(new_current_goal_state)

  def updateCurrentGoalState(self, dt):
    return self.goal_sys.updateCurrentGoalState(dt)

  def getSteadyStateGoalPosition(self):
    return self.goal_sys.getSteadyStateGoalPosition()

  def setSteadyStateGoalPosition(self, new_G):
    return self.goal_sys.setSteadyStateGoalPosition(new_G)

  def setCouplingTermUsagePerDimensions(
      self, is_using_coupling_term_at_dimension_init):
    assert (
        len(is_using_coupling_term_at_dimension_init) == self.dmp_num_dimensions
    ), "len(is_using_coupling_term_at_dimension_init)=" + str(
        len(is_using_coupling_term_at_dimension_init)
    ) + " is mis-matched with self.dmp_num_dimensions=" + str(
        self.dmp_num_dimensions) + "!"
    self.is_using_coupling_term_at_dimension = is_using_coupling_term_at_dimension_init
    return None
